contract/models/contract_line.py

```diff
@@ -542,23 +577,9 @@ def _get_period_to_invoice(
             if last_date_invoiced
             else self.date_start
         )
-        if self.recurring_rule_type == 'monthlylastday':
-            last_date_invoiced = recurring_next_date
-        else:
-            if self.recurring_invoicing_type == 'pre-paid':
-                last_date_invoiced = (
-                    recurring_next_date
-                    + self.get_relative_delta(
-                        self.recurring_rule_type, self.recurring_interval
-                    )
-                    - relativedelta(days=1)
-                )
-            else:
-                last_date_invoiced = recurring_next_date - relativedelta(
-                    days=1
-                )
+        last_date_invoiced = self.next_period_end_date
         if stop_at_date_end:
-            if self.date_end and self.date_end < last_date_invoiced:
+            if not self.next_period_end_date and self.date_end:
                 last_date_invoiced = self.date_end
         return first_date_invoiced, last_date_invoiced, recurring_next_date
```

Comment:
When you have done `_update_next_period` this `_get_period_to_invoice` can become very dumb: `return next_period_start_date, next_period_end_date`